language/diem-vm/src/diem_vm.rs

```diff
@@ -468,54 +469,58 @@ impl<'a> DiemVMInternals<'a> {
     }
 }
 
-pub fn txn_effects_to_writeset_and_events_cached<C: AccessPathCache>(
+pub fn convert_changeset_and_events_cached<C: AccessPathCache>(
     ap_cache: &mut C,
-    effects: TransactionEffects,
+    changeset: MoveChangeSet,
+    events: Vec<MoveEvent>,
 ) -> Result<(WriteSet, Vec<ContractEvent>), VMStatus> {
     // TODO: Cache access path computations if necessary.
     let mut ops = vec![];
 
-    for (addr, vals) in effects.resources {
-        for (struct_tag, val_opt) in vals {
+    for (addr, account_changeset) in changeset.accounts {
+        for (struct_tag, blob_opt) in account_changeset.resources {
             let ap = ap_cache.get_resource_path(addr, struct_tag);
-            let op = match val_opt {
+            let op = match blob_opt {
                 None => WriteOp::Deletion,
-                Some((ty_layout, val)) => {
-                    let blob = val.simple_serialize(&ty_layout).ok_or(VMStatus::Error(
-                        StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
-                    ))?;
-
-                    WriteOp::Value(blob)
-                }
+                Some(blob) => WriteOp::Value(blob),
             };
             ops.push((ap, op))
         }
-    }
 
-    for (module_id, blob) in effects.modules {
-        ops.push((ap_cache.get_module_path(module_id), WriteOp::Value(blob)))
+        for (name, blob_opt) in account_changeset.modules {
+            let ap = ap_cache.get_module_path(ModuleId::new(addr, name));
+            let op = match blob_opt {
+                None => WriteOp::Deletion,
+                Some(blob) => WriteOp::Value(blob),
+            };
+
+            ops.push((ap, op))
+        }
     }
 
     let ws = WriteSetMut::new(ops)
         .freeze()
```

Comment:
This is another thing we should fix in a follow-up: `freeze` actually can't fail, we should get rid of the `Result` return value.